### ISSUE TITLE
Circular typedef references causes incorrect error

### ISSUE THREAD

**Issue by sandersn** [microsoft/TypeScript#34809]
```js
// @filename: test.js
const mod1 = require('./mod1')
/** @typedef {number} CompilerStatus */
/** @type {CompilerStatus[]} */
var compilerStatus;
mod1;
module.exports = class C {
}

// @filename: mod1.js
/** @type {import("./test")} */
var megaman;
module.exports = megaman
```

Expected: No error, or maybe a circular reference error somewhere.

Actual: Error on `@typedef {number} CompilerStatus`: "Duplicate identifier 'CompilerStatus'.".

Bonus: in test.js, if you export `module.exports = {}` you get a stack overflow from the language service. I couldn't get this from the compiler, but I didn't try too hard.

**Comment by sandersn** [comment]
From webpack: MultiCompiler.js and MultiWatching.js

**Comment by sandersn** [comment]
Doesn't repro anymore, but webpack still has the same failure. Looks like I need to fix the repro.

**Comment by sandersn** [comment]
I haven't got a repro but I've narrowed it down to a symbol-merging bug~~~~~!

I knew we shouldn't have removed that assert in mergeSymbolTable.

**Comment by sandersn** [comment]
Looks like the first failure was on https://github.com/microsoft/TypeScript/pull/34718/files on 1:41 PM October 24.

**Comment by sandersn** [comment]
Looks like it was caused by https://github.com/microsoft/TypeScript/pull/34712

**Comment by sandersn** [comment]
Here's a standalone repro:

```json
{
  "compilerOptions": {
    "allowJs": true,
    "checkJs": true,
    "noEmit": true
  }
}
```

and

```js
// @Filename: MW.js
/** @typedef {import("./MC")} MC */

class MW {
  /**
   * @param {MC} compiler the compiler
   */
  constructor(compiler) {
    this.compiler = compiler;
  }
}

module.exports = MW;

// @Filename: MC.js
const MW = require("./MW");

/** @typedef {number} Cictema */

module.exports = class MC {
  watch() {
    return new MW(this);
  }
};
```

Gives the following compile error:

```
MC.js:3:23 - error TS2300: Duplicate identifier 'Cictema'.

3 /** @typedef {number} Cictema */
                        ~~~~~~~

  MC.js:3:5
    3 /** @typedef {number} Cictema */
          ~~~~~~~~~~~~~~~~~~~~~~~~~
    'Cictema' was also declared here.
  MC.js:3:5
    3 /** @typedef {number} Cictema */
          ~~~~~~~~~~~~~~~~~~~~~~~~~
    and here.


Found 1 error.
```

The error appears in the editor too. I'm not really sure what's different from the repro I posted originally.
